update vignette on making custom teal module (#1400)
insightsengineering/teal: vignettes/creating-custom-modules.Rmd

## Creating the Custom `teal` Module Function

The `teal::module()` function allows you to encapsulate your UI and server logic into a `teal` module, making it reusable and ready to integrate into any `teal` application.

By setting `datanames = "all"`, you give the module access to all datasets specified in the `teal_data` object.

```r
# Custom histogram module creation
create_histogram_module <- function(label = "Histogram Module") {
  teal::module(
    label = label,
    ui = histogram_module_ui,
    server = histogram_module_server,
    datanames = "all"
  )
}
```

## Integrating the Custom `teal` Module into a `teal` App

With the custom `teal` module set up, it can now be integrated into a `teal` app. 
We’ll use `init()` from `teal` to specify the datasets and modules used in the app, then run the app to test the newly created module.

```r
library(teal)

# Define datasets in `teal_data`
data_obj <- teal_data(
  iris = iris,
  mtcars = mtcars
)

# Initialize the teal app
app <- init(
  data = data_obj,
  modules = modules(create_histogram_module())
)

# Run the app
shiny::shinyApp(ui = app$ui, server = app$server)

> ```

Reviewer: WDYT about adding a Shinylive compontent in here - right after the final app?

More general issue: https://github.com/insightsengineering/teal/issues/1404. We can do it now in this PR or wait for this.

Author: I did attempt to add the link as you suggested but it's not working because we haven't release the updated version of teal.data (using `names` instead of `datanames` when pulling the names of data objects in teal_data object). I think we should wait.

The `teal` framework provides a large catalog of plug-in-ready analysis modules to be incorporated into `teal` applications.
However, it is also possible to create your own modules using the `module` function.

In this guide, we will convert a simple histogram formula into a robust `teal` module:
```r

> hist(dataset[, selected], las = 1)

Author: Good idea, but maybe avoid using the `input_` naming convention. The goal is to replace `dataset` with `input_dataset` when writing this in the server logic.

> hist(dataset[, selected], las = 1)

Author: Update was done in c508c133cbc1c128a9abbf530b15ec42cff8d338

> hist(dataset[, selected], las = 1)

Reviewer 2: I was checking this and realized the variables names under the section "Understanding the Inputs and Requirements" doesn't match these (it has `selected` instead of `vars` or `variable`).

> hist(dataset[, selected], las = 1)

Author: nice catch!

> hist(dataset[, selected], las = 1)

Author: Done in 5617508983db5686177b3260a469fdbb71a0e6fb

This `data` argument holds the reactive `teal_data` object, which contains your datasets and any filters applied. Including `data` ensures:

- The server function receives a reactive version of `teal_data`, allowing it to automatically respond to changes.
- The server can access the filtered datasets directly.

The correct function definition for the server function is:

```r
histogram_module_server <- function(id, data) {
  moduleServer(id, function(input, output, session) {
    # Server logic goes here
  })
}
```

### Understanding `teal_data` as a Reactive Object in Server Logic

When used in the server logic of a `teal` module, the `teal_data` object becomes a **reactive data container**.
This means that to access its contents, you need to call it like a function, using parentheses: `data()`.

This syntax triggers reactivity, ensuring that the data within `teal_data` stays up-to-date with any filters or changes applied elsewhere in the application.

> **Note**: The `teal_data` object behaves as a reactive data container only when used within the server logic. If accessed outside of the server, it will not be reactive.

> ### Using `datanames()` to Access Dataset Names in `teal_data` object

Author: I will need to change this to just using `names()` once `datanames()` is deprecated.

Author: Done with 5617508983db5686177b3260a469fdbb71a0e6fb

The `teal` framework provides a large catalog of plug-in-ready analysis modules to be incorporated into `teal` applications.

> However, it is also possible to create your own modules using the `module` function.

Author: I'm opposed to it. 

What do you think about this sentence?

_The teal framework provides a large catalog of plug-in-ready analysis modules that can be incorporated into teal applications. However, it is also possible to create your own modules using the module function, which leverages Shiny modules. Each custom teal module is built as a Shiny module, combining Shiny's reactive capabilities with modularized UI and server logic to encapsulate functionality. This design enables a structured and reusable approach to creating interactive components that integrate seamlessly within the teal ecosystem._

> However, it is also possible to create your own modules using the `module` function.

Reviewer 3: Awesome!

> However, it is also possible to create your own modules using the `module` function.

Author: Done with 5617508983db5686177b3260a469fdbb71a0e6fb

In this guide, we will use the simple histogram below as an example, and demonstrate how to convert this histogram function into a robust `teal` module step-by-step:
```r
my_plot <- hist(
  dataset[[vars]],
  las = 1,
  main = paste("Histogram of", vars),
  xlab = vars,
  col = "lightblue",
  border = "black"
)

> ```

Author: Done with 5617508983db5686177b3260a469fdbb71a0e6fb

> This module will allow users to dynamically select datasets and variables to create histograms within a `teal` application.

Author: I checked the rest of the sections do have an empty line before/after code chunk and after title sub-section as you mentioned. I may have missed this one. Thanks for catching this!

> This module will allow users to dynamically select datasets and variables to create histograms within a `teal` application.

Author: Done with 5617508983db5686177b3260a469fdbb71a0e6fb

> We will cover best practices, including setting up dynamic inputs, structuring server logic, and using the `teal_data` object to ensure reactivity and reproducibility.

Author: I missed this on my initial review but I agree with the suggestion. I'll make this update.

- **`plotOutput` for Histogram**: Displays the histogram once both dataset and variable inputs are selected.
- **`verbatimTextOutput` for Code**: Automatically displays code that generated the plot based on user input.

Here’s the code for the `histogram_module_ui` function:

```r
# UI function for the custom histogram module
histogram_module_ui <- function(id) {
  ns <- shiny::NS(id)
  shiny::tagList(
    shiny::selectInput(ns("dataset"), "Select Dataset", choices = NULL),
    shiny::selectInput(ns("variable"), "Select Variable", choices = NULL),
    shiny::plotOutput(ns("histogram_plot")),
    shiny::verbatimTextOutput(ns("plot_code"))  # To display the reactive plot code
  )
}
```

## Setting Up the `teal` Module Server

The server function is where the main logic of a `teal` module is handled. 
For our histogram module, the server function will handle user interactions and manage the reactive `teal_data` object, which allows the module to dynamically respond to user inputs.

### Passing the `data` Argument to the Server Function

> 

Author: I'm a bit on the fence about this because I’m hoping users have already reviewed the "Getting Started with teal" vignette, which covers `teal_data`.
https://insightsengineering.github.io/teal/latest-tag/articles/getting-started-with-teal.html#application-data

<br />
<br />

> Link that was in the previous documentation https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data-reproducibility.html

The link provided in the previous documentation made sense because the sentence before it discussed code reproducibility. Including it here without any lead-in context doesn’t quite fit.

If we think it’s needed, we could close this sub-category with a reminder sentence like this:

_If you need a refresher on the teal_data object, please visit the [teal.data package documentation](https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data.html)._

What do you think?

> 

Reviewer 3: > If you need a refresher on the teal_data object, please visit the [teal.data package documentation](https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data.html).

:+1:

> 

Author: Done with 5617508983db5686177b3260a469fdbb71a0e6fb

2. **Rendering the Plot**:
    - `output$histogram_plot` uses `renderPlot()` to display the plot stored in `new_data` by referencing `result()[["my_plot"]]`.

3. **Plot Code Display**:
    - The `output$plot_code` render function displays the dynamically generated code using `teal.code::get_code(result())`, allowing users to see the exact code used to generate the plot reactively.

## Creating the Custom `teal` Module Function

The `teal::module()` function allows you to encapsulate your UI and server logic into a `teal` module, making it reusable and ready to integrate into any `teal` application.

By setting `datanames = "all"`, you give the module access to all datasets specified in the `teal_data` object.

```r
# Custom histogram module creation
create_histogram_module <- function(label = "Histogram Module") {
  teal::module(
    label = label,
    ui = histogram_module_ui,
    server = histogram_module_server,
    datanames = "all"
  )
}
```

## Integrating the Custom `teal` Module into a `teal` App

With the custom `teal` module set up, it can now be integrated into a `teal` app. 

> We’ll use `teal::init()` to specify the datasets and modules used in the app, then run the app to test functionality.

Reviewer 3: ```suggestion
We’ll use `init()` from `teal` to specify the datasets and modules used in the app, then run the app to test the newly created module.
```

```r
library(teal)

# Define datasets in `teal_data`

> data_obj <- teal.data::teal_data(
>   iris = iris,
>   mtcars = mtcars
> )
> 
> # Initialize the teal app
> app <- teal::init(
>   data = data_obj,
>   modules = teal::modules(create_histogram_module())
> )

Reviewer 3: I think we can ditch the prefix, given that the library call exists (that is what we do in other vignettes)

```suggestion
data_obj <- teal_data(
  iris = iris,
  mtcars = mtcars
)

# Initialize the teal app
app <- init(
  data = data_obj,
  modules = modules(create_histogram_module())
)
```

# Run the app

> if (interactive()) {

Author: Good point. I'll remove it.

> if (interactive()) {

Author: Done with 5617508983db5686177b3260a469fdbb71a0e6fb

  shiny::shinyApp(ui = app$ui, server = app$server)
}
```

**Congratulations! You just created a custom teal module and used it in a teal app!**

This setup provides a fully dynamic, user-controlled `teal` module that allows for interactive data exploration and code visibility, enhancing both usability and transparency.

> ## What's next?

Reviewer 3: (extra new line for markdown)

```suggestion
## What's next?

```

Now that you’ve mastered the essentials of building and integrating modules in `teal`, you’re ready to explore more advanced features. 
`teal` offers a wide range of capabilities to enhance your module’s functionality and user experience. 

> ### Adding reporting to a module

Reviewer 3: (extra new line for markdown)

```suggestion
### Adding reporting to a module

```
